mlx/backend/metal/kernels/fence.metal

```diff
@@ -0,0 +1,52 @@
+// Copyright © 2024 Apple Inc.
+
+#pragma METAL internals : enable
+
+#ifndef __METAL_MEMORY_SCOPE_SYSTEM__
+#define __METAL_MEMORY_SCOPE_SYSTEM__ 3
+namespace metal {
+constexpr constant metal::thread_scope thread_scope_system =
+    static_cast<thread_scope>(__METAL_MEMORY_SCOPE_SYSTEM__);
+}
+#endif
+
+#include <metal_atomic>
+
+[[kernel]] void input_coherent(
```

Comment:
What does the `input_coherent` do? Does it mean to ensure the data gets fully transferred from cpu to gpu?

Reply:
It's to ensure the data is safe to be read by the CPU.

Reply:
Is this a metal specific thing?

I have a weird flaky test in cuda backend that an array loaded from cpu sometimes ends up with wrong value in gpu:
https://github.com/ml-explore/mlx/blob/9daa6b003f548569e9186502d5acb9b74b91bcbe/tests/load_tests.cpp#L36

I wonder if I need to do something similar here, though it is completely different hardware.

Reply:
It's very metal specific. In fact.. most of the time this kerne is not used. It's only enabled for an experimental polling based synchronization when you set `MLX_FAST_METAL_SYNCH=1`. 

If you are using standard cuda synchronization primitives you should need anything like that.

Reply:
Cuda does not have a standard synchronization primitive for signaling a gpu wait from cpu, I was using `cuda::atomic` but its `wait/notify` APIs had some problems with CUDA 11 (https://github.com/ml-explore/mlx/issues/2137) so I'm currently using busy wait for fence which is almost identical to metal backend's fast fence.

Reply:
> so I'm currently using busy wait for fence which is almost identical to metal backend's fast fence.

Oh right. Well in that case.. we are in pretty uncharted territory. Do you know of examples that have used a similar concept in CUDA? That would be a good place to look and see if they have to do anything fancy to ensure the input safe to read. 

In some Metal we need this because even if the fence value is written by the GPU kernel (which certainly implies and the input is done being computed) it can still be in cache and not necessarily "coherent" with the rest of the device (and system in unified memory).

Reply:
No I couldn't find any open source code synchronizes gpu with cpu work, and there is actually no framework using cuda's unified memory APIs. I also searched nvidia forum and there was probably 2 posts related to this topic in past 10 years.